### ISSUE TITLE
[REVIEW]: MaDaTS: Managing Data on Tiered Storage for Scientific Workflows

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#830]
**Submitting author:** @dghoshal-lbl (<a href="http://orcid.org/0000-0002-6819-6949">Devarshi Ghoshal</a>)
**Repository:** <a href="https://github.com/dghoshal-lbl/madats" target ="_blank">https://github.com/dghoshal-lbl/madats</a>
**Version:** v1.1.2
**Editor:** @danielskatz
**Reviewer:** @ian-taylor, @gflofst
**Archive:** <a href="https://doi.org/10.5281/zenodo.1439288" target="_blank">10.5281/zenodo.1439288</a>

## Status

[![status](http://joss.theoj.org/papers/0ec885baba24844be3c89eb532ecc9b0/status.svg)](http://joss.theoj.org/papers/0ec885baba24844be3c89eb532ecc9b0)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/0ec885baba24844be3c89eb532ecc9b0"><img src="http://joss.theoj.org/papers/0ec885baba24844be3c89eb532ecc9b0/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/0ec885baba24844be3c89eb532ecc9b0/status.svg)](http://joss.theoj.org/papers/0ec885baba24844be3c89eb532ecc9b0)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/dghoshal-lbl/madats" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@ian-taylor & @gflofst, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @danielskatz know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/dghoshal-lbl/madats">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v1.1.2)?
- [x] **Authorship:** Has the submitting author (dghoshal-lbl) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/dghoshal-lbl/madats">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v1.1.2)?
- [x] **Authorship:** Has the submitting author (dghoshal-lbl) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks. @ian-taylor, it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00830/joss.00830/10.21105.joss.00830.pdf)

**Comment by danielskatz** [comment]
👋  @ian-taylor & @gflofst - please go ahead and start your reviews

See the first comment in the issue for details on how to do the review, and  https://joss.readthedocs.io/en/latest/reviewer_guidelines.html and https://joss.readthedocs.io/en/latest/review_criteria.html for more details. In brief, what I'm looking for is you both to check off the 18 items on your checklists.

If you have problems, feel free to comment here or email me.

Finally, a note that @ian-taylor will not be able to do his review until the 2nd week of August.

**Comment by danielskatz** [comment]
👋 @ian-taylor & @gflofst - any update on your reviews?  I don't see any checkmarks yet...

**Comment by gflofst** [comment]
Today.

On Jul 30, 2018, at 10:35 AM, Daniel S. Katz <notifications@github.com<mailto:notifications@github.com>> wrote:


👋 @ian-taylor<https://github.com/ian-taylor> & @gflofst<https://github.com/gflofst> - any update on your reviews? I don't see any checkmarks yet...

**Comment by gflofst** [comment]
The license is not a standard license, but a University of California/DOE license. It is very straightforward reminiscent of the MIT license.

**Comment by danielskatz** [comment]
Thanks @gflofst 

👋 @dghoshal-lbl - Note that JOSS requires using [an OSI-approved license](https://opensource.org/licenses/alphabetical) - which is this?

**Comment by gflofst** [comment]
It has no text citing it as a particular license. Instead, it lists that it is approved by Uni. Cal. and LBL (and DOE). The license text is basically MIT (do what you want, cite this code, we are not responsible for what you do or what happens because of use).

**Comment by dghoshal-lbl** [comment]
@danielskatz @gflofst It is a modified BSD license.

**Comment by gflofst** [comment]
In the setup.py, it says BSD. How exacting should that match be to be acceptable?

**Comment by danielskatz** [comment]
Modified in any way is not acceptable.

To be published by JOSS, the code needs to have an OSI-approved license.

**Comment by dghoshal-lbl** [comment]
There is an additional paragraph at the end to the original BSD license (due to lab policies).

**Comment by dghoshal-lbl** [comment]
Is it mandatory to have the exact same license? Previously one of the papers was accepted because the license is still BSD, with an additional paragraph.

**Comment by danielskatz** [comment]
Yes, it is mandatory.  Can I check the process for the other paper?  Which was it?

**Comment by gflofst** [comment]
Community guidelines are either missing or not obvious

**Comment by dghoshal-lbl** [comment]
@danielskatz 
i) This is the paper: https://joss.theoj.org/papers/f68f92a740c6e42365025be9a9efed1b
ii) Software package: https://zenodo.org/record/290299#.W1-isdhKjOQ (Please note license on the right side, which says --  BSD 3-Clause "New" or "Revised" License (BSD-3-Clause))
iii) And the review comments: https://bitbucket.org/dghoshal/frieda/issues/1/joss-review-comments

**Comment by dghoshal-lbl** [comment]
@gflofst A sample storage.yaml file is created when setup.py is run. Did the build process fail?

**Comment by danielskatz** [comment]
👋 @arfon  - please weigh in on the license discussion for this submission, as well as for https://doi.org/10.21105/joss.00164

**Comment by arfon** [comment]
> Is it mandatory to have the exact same license? Previously one of the papers was accepted because the license is still BSD, with an additional paragraph.

I think this statement is contradictory. The whole point of having a standard set of approved and vetted open source licenses is that they are used verbatim _without modification_.

This submission should be considered blocked by this issue and I think the JOSS editorial board needs to discuss the status of https://doi.org/10.21105/joss.00164 as the non-standard license is problematic and something we should have caught (and likely would going forward with this https://github.com/openjournals/whedon-api/pull/35).

**Comment by gflofst** [comment]
@dghoshal-lbl Here is the output when I try to run setup. It looks like I should have created a directory, but this isn't specified in the instructions. Please review and adjust the instructions accordingly.
[setup.txt](https://github.com/openjournals/joss-reviews/files/2246568/setup.txt)

**Comment by dghoshal-lbl** [comment]
@gflofst There is no need to create any additional directory. The config/ directory is already present in the top-level directory of the source tree. 
Unfortunately I am unable to reproduce the error on my side. I need some more information.
- Is this where you have checked out the source (including setup.py): /Users/gflofst/build/joss/run?
- Have you manually configured MADATS_HOME environment variable?

**Comment by gflofst** [comment]
I did the clone into joss so the top level directory is
/Users/gflofst/build/joss/madats

I do not have MADATS_HOME set in this terminal window. While the instructions after the two steps say to set it up, it should be a step in the instructions in order too. When I do set it to the above directory, it works. Having to set it to generate the file that will set it doesn't make sense.

**Comment by dghoshal-lbl** [comment]
You do not need to set MADATS_HOME manually. I will fix the instructions in the readme if that is not clear enough. However, if you run setup from the top level directory "/Users/gflofst/build/joss/madats" (where setup.py is), it should create the required files. 
I was wondering why the setup log shows "/Users/gflofst/build/joss/run" as MADATS_HOME. What is this joss/run/ directory, if code is in joss/madats/?

**Comment by gflofst** [comment]
code is in joss/madats and that is where I run setup.py from. The only way I can get it to work properly is if I set MADATS_HOME manually before running setup.py.

**Comment by dghoshal-lbl** [comment]
That shouldn't happen. Can you please open up a new terminal and let me know if "echo $MADATS_HOME" prints anything (without you manually setting it up)? If it doesn't print anything, can you please rerun "python setup.py install" and send me the output of the build process?

I tried in a couple of different environments and unable to reproduce the error. Can you please check if there is a "storage.yaml" file present in config/ directory at the end of the build process?

**Comment by gflofst** [comment]
Oddly, I closed my terminal so that I cleared out any env vars I set manually and cleaned the repo clone and now it works. I am in the same directory and everything.

I'll consider this pass now. Not sure what my problem was. If you could clarify the readme instructions for setup to just have the source MADATS_HOME as the third step instead of the next paragraph, that would help.

I am slightly concerned that this is bash centric, but that might be my old timer personality speaking. Bash pretty well has taken over, but it is not a 100% solution. A note to adjust for other shells would make it 100%, but I am not sure how necessary that is these days.

**Comment by gflofst** [comment]
At this point, the only two pieces of contention are the license (out of my hands) and the need for a "how to contribute" set of instructions. Everything else looks good.

**Comment by dghoshal-lbl** [comment]
Thanks @gflofst. Here are the action items for me to fix:
i) Fix the readme to have "source MADATS_HOME" as the third step.
ii) A set of instructions on "how to contribute".
iii) Since the bash-centric command is only for setting up MADATS_HOME, I will add a few lines in the readme to setup MADATS_HOME on other environments.

Let me know if there are any other changes that you think are necessary. I will fix these and let you know.

I hope the licensing issue will be resolved soon.

**Comment by gflofst** [comment]
That is what I see. Sorry for the troubles.

**Comment by danielskatz** [comment]
We have an editors call on Friday where the licensing issue will be discussed - I think it's going to be blocking problem, so @dghoshal-lbl, you might want to think about what you are able to do or will choose to do if so.

**Comment by gflofst** [comment]
@danielskatz, any update?

**Comment by danielskatz** [comment]
@gflofst & @ian-taylor - please put this on hold on now.  We cannot proceed with the current license, so either the license will be changed, or we will remove this from the JOSS queue.  We've asked the author to let us know what he can do in the next three weeks, with the understanding that employer policies may not permit a satisfactory resolution in that time - if so, the paper will likely be withdrawn until it meets the JOSS policies at which time it could be resubmitted, and we could use the reviewing work that has already been done as a starting point.

**Comment by dghoshal-lbl** [comment]
@gflofst I have updated the repository with the changes you proposed. I have updated the readme and added a 'CONTRIBUTING' file. The license is also changed to OSI approved 3-clause BSD license. 

Please let me know if everything looks okay and if it needs any more edits/modifications.

**Comment by danielskatz** [comment]
Thanks @dghoshal-lbl - I'm glad you were able to do that so quickly.

@gflofst - back to you to complete your review

and @ian-taylor, for you to start yours.

**Comment by gflofst** [comment]
@dghoshal-lbl , the checklist for the contributions looks for 3 things:
Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support
I see the first one, but not the second 2.

**Comment by dghoshal-lbl** [comment]
@gflofst The issues can be submitted through Github issues for the repository. Do you want me to add the information in the "CONTRIBUTING" file?

**Comment by gflofst** [comment]
@dghoshal-lbl I agree it is a bit redundant since it is expected that if it is on github that you'll respond to anything reported there, but I think it would be best to list it in the file just to be clear. For me, I'd list to use that or if it proves unresponsive (more than 1 week before some response), to contact me via email. I don't think that is necessary for this journal checklist though.

**Comment by dghoshal-lbl** [comment]
@gflofst Ok, makes sense. I will add a couple of lines about it.

**Comment by danielskatz** [comment]
👋 @ian-taylor - any update on your status on this review?

**Comment by ian-taylor** [comment]
ok, here is what I did to start from scratch using python 2.7 and pip version.x 18:

git clone https://github.com/dghoshal-lbl/madats.git
cd madats
virtualenv madats
source madats/bin/activate
pip install --upgrade pip
(and pip -V
pip 18.0 from /Users/ian/Sites/madats/madats/madats/lib/python2.7/site-packages/pip (python 2.7)
)
pip install -r requirements.txt
python setup.py install
source MADATS_HOME
py.test tests/test_madats.py

And I got:

============================================= test session starts ==============================================
platform darwin -- Python 2.7.10, pytest-3.4.0, py-1.5.4, pluggy-0.6.0
rootdir: /Users/ian/Sites/madats/madats, inifile:
collected 0 items / 1 errors                                                                                   

==================================================== ERRORS ====================================================
____________________________________ ERROR collecting tests/test_madats.py _____________________________________
ImportError while importing test module '/Users/ian/Sites/madats/madats/tests/test_madats.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/test_madats.py:19: in <module>
    import madats
madats/lib/python2.7/site-packages/MaDaTS-1.1.2-py2.7.egg/madats/__init__.py:1: in <module>
    from madats.core.vds import VirtualDataSpace, VirtualDataObject, Task
madats/lib/python2.7/site-packages/MaDaTS-1.1.2-py2.7.egg/madats/core/vds.py:24: in <module>
    from scandir import scandir
E   ImportError: No module named scandir
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=========================================== 1 error in 0.29 seconds ============================================


-- So, should I assume scandir should be available?, did I do something else wrong? or should this be added to the install requirements.txt file and/or instructions?

**Comment by danielskatz** [comment]
thanks for getting started @ian-taylor  - also, please check off items on your part of the checklist at the top...

**Comment by ian-taylor** [comment]
yes, I got that ....

**Comment by danielskatz** [comment]
I was hoping you would check the COI and code of conduct ones before you started on other things :)

**Comment by ian-taylor** [comment]
Ok... on to more important things

**Comment by dghoshal-lbl** [comment]
@ian-taylor Not sure how I missed that. I will add scandir to requirements.txt

**Comment by dghoshal-lbl** [comment]
@ian-taylor I have fixed requirements.txt. Let me know if you run into any other problem.

**Comment by ian-taylor** [comment]
Its quite hard to get going on this without creating your own workflow from scratch.  None of the examples in the examples directory work as is.  To get any of these to work, as stated in each example you need to "Update config/storage.yaml with the appropriate storage tiers and specify the datadir to make this example work"

Now, the madats_workflow.yaml is generic and consists of two tasks.  The description in the UserGuide  "MaDaTS Workflow Description" provides a list of what the options mean but there is no overview of how the workflow works. 

For example, one would expect a way of defining dependencies between tasks if this is a workflow system. How can that be achieved? or is this more of a batch oriented system i.e. execute task1 then task2, etc ?  

What would be useful is a fully operating example that a programmer can run to try this system out, even if it is simple.    Also, it would be helpful to provide 2 or 3 other functional workflows that help walk a programmer through the process of building simple batch flows and then more complex workflows.

**Comment by dghoshal-lbl** [comment]
@ian-taylor The example section in Readme lists the steps to create a workflow of your own. But I think you are asking for a script or a sample workflow that you can tweak to build your own workflow. Is that correct?

**Comment by dghoshal-lbl** [comment]
@ian-taylor I think I have a few examples in the source tree. I will check those and see if I can add a few lines in the readme/user-guide on how to use those and build your own workflows using the examples as a reference. Hopefully that will help.

**Comment by danielskatz** [comment]
@ian-taylor  - how are things looking to you?

**Comment by dghoshal-lbl** [comment]
@danielskatz @ian-taylor There are a few examples in the examples/ directory in the source tree. I am adding more information about these examples in the user guide so that users can use these as reference.

**Comment by ian-taylor** [comment]
Can you let me know when the documentation is complete so I can try and verify ?

**Comment by dghoshal-lbl** [comment]
@ian-taylor I have added Chapter-3 in UserGuide.pdf to explain some examples of using MaDaTS. Let me know if it addresses what you asked for.

**Comment by dghoshal-lbl** [comment]
@ian-taylor Any update on the review?

**Comment by danielskatz** [comment]
👋  @gflofst - what about the community guidelines checklist item? Can you check this off, or let us you what else you think is needed?

**Comment by danielskatz** [comment]
@dghoshal-lbl - please archive the current version somewhere (e.g. Zenodo) and let me know the DOI of the archive.

**Comment by dghoshal-lbl** [comment]
@danielskatz - here's the DOI: 10.5281/zenodo.1439288

**Comment by arfon** [comment]
@ian-taylor, @gflofst - many thanks for your reviews here and to @danielskatz for editing this submission ✨ 

@dghoshal-lbl - your paper is now accepted into JOSS and your DOI is https://doi.org/10.21105/joss.00830 :zap: :rocket: :boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.00830/status.svg)](https://doi.org/10.21105/joss.00830)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.00830">
  <img src="http://joss.theoj.org/papers/10.21105/joss.00830/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.00830/status.svg
   :target: https://doi.org/10.21105/joss.00830
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.00830/status.svg)](https://doi.org/10.21105/joss.00830)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3Nzk=

**Comment by dghoshal-lbl** [comment]
@arfon @danielskatz @gflofst @ian-taylor Thank you all for reviewing and getting the paper published!
